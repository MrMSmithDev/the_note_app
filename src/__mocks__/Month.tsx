import { MonthProps } from '@components/Month/Month';

export default function ({ month, year }: MonthProps) {
  return (
    <div data-testid="mock-month">
      {month.abbr}
    </div>
  );
}
