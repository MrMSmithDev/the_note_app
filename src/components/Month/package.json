{
  "name": "month",
  "version": "1.0.0",
  "main": "Month.tsx",
  "dependencies": {
    "react": "^19.0.0"
  }
}
