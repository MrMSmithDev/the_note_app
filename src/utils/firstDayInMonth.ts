export default function firstDayInMonth(month: number, year: number) {
  return new Date(year, month - 1, 1).getUTCDay();
}
