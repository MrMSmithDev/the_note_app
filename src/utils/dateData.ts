import Month from 'src/types/MonthType';

const dayData: string[] = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];

const monthsData: Month[] = [
  { abbr: 'JAN', num: 1 },
  { abbr: 'FEB', num: 2 },
  { abbr: 'MAR', num: 3 },
  { abbr: 'APR', num: 4 },
  { abbr: 'MAY', num: 5 },
  { abbr: 'JUN', num: 6 },
  { abbr: 'JUL', num: 7 },
  { abbr: 'AUG', num: 8 },
  { abbr: 'SEP', num: 9 },
  { abbr: 'OCT', num: 10 },
  { abbr: 'NOV', num: 11 },
  { abbr: 'DEC', num: 12 },
];

export { dayData, monthsData };
