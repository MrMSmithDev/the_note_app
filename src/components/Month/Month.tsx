import Date from '@components/Date';
import { dayData } from '@utils/dateData';
import daysInMonth from '@utils/daysInMonth';
import firstDayInMonth from '@utils/firstDayInMonth';
import { useEffect, useState } from 'react';
import MonthType from 'src/types/MonthType';

export interface MonthProps {
  month: MonthType;
  year: number;
}

const Month: React.FC<MonthProps> = ({ month, year }) => {
  const [days, setDays] = useState<{ date: number }[]>([{ date: 1 }]);
  const [leadingSpace, setLeadingSpace] = useState<number>(0);

  useEffect(() => {
    const daysAmount = daysInMonth(month.num, year);
    const firstDay = firstDayInMonth(month.num, year);

    setLeadingSpace(firstDay);

    const daysArr = Array.from({ length: daysAmount }, (_, i) => ({
      date: i + 1,
    }));
    setDays(daysArr);
  }, [year]);

  return (
    <div className="flex flex-col mb-2 md:mb-0">
      <h3 className="text-center font-bold font-quicksand text-md md:text-sm select-none dark:text-gray-100">
        {month.abbr}
      </h3>
      <div className="grid grid-cols-7 gap-2 my-3 font-quicksand">
        {dayData.map((day, i) => (
          <span
            key={`${day}-${i}`}
            className="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
          >
            {day}
          </span>
        ))}
        <div
          role="presentation"
          style={{ gridColumn: `span ${leadingSpace}` }}
        />
        {days.map((day: { date: number }) => (
          <div key={day.date} className="font-quicksand text-center">
            <Date date={day.date} month={month} year={year} />
          </div>
        ))}
      </div>
    </div>
  );
};

export default Month;
