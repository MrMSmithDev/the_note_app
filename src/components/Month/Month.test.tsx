import { render, screen } from '@testing-library/react';
import Month from './Month';
import { dayData } from '@utils/dateData';
import daysInMonth from '@utils/daysInMonth';
import firstDayInMonth from '@utils/firstDayInMonth';

jest.mock('@components/Date', () => require('@__mocks__/Date'));

jest.mock('@utils/daysInMonth', () => jest.fn());
jest.mock('@utils/firstDayInMonth', () => jest.fn());

describe('Month component', () => {
  const mockMonth = { num: 2, abbr: 'FEB' }; // February
  const mockYear = 2025;

  beforeEach(() => {
    jest.clearAllMocks();

    (daysInMonth as jest.Mock).mockReturnValue(28);
    (firstDayInMonth as jest.Mock).mockReturnValue(3);
  });

  it('renders correctly to match snapshot', () => {
    const { asFragment } = render(<Month month={mockMonth} year={mockYear} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correct month abbreviation', () => {
    render(<Month month={mockMonth} year={mockYear} />);

    expect(screen.getByText('FEB')).toBeInTheDocument();
  });

  it('renders the correct number of days for the month', () => {
    render(<Month month={mockMonth} year={mockYear} />);

    const days = screen.getAllByTestId('mock-date');
    expect(days.length).toBe(28);
    expect(days[0]).toHaveTextContent('1');
    expect(days[27]).toHaveTextContent('28');
  });

  it('creates correct leadings spaces in grid', () => {
    render(<Month month={mockMonth} year={mockYear} />);

    const leadingSpaceDiv = screen.getByRole('presentation', { hidden: true });
    expect(leadingSpaceDiv).toHaveStyle('grid-column: span 3');
  });

  it('renders the weekday labels correctly', () => {
    render(<Month month={mockMonth} year={mockYear} />);

    dayData.forEach((day) => {
      const elements = screen.getAllByText(day, { exact: true });
      expect(elements.length).toBeGreaterThan(0);
    });
  });

  it('updates when the year changes', () => {
    const { rerender } = render(<Month month={mockMonth} year={2024} />);
    expect(screen.getAllByTestId('mock-date').length).toBe(28);

    (daysInMonth as jest.Mock).mockReturnValue(31);
    rerender(<Month month={{ abbr: 'MAR', num: 3 }} year={2025} />);

    expect(screen.getByText('MAR')).toBeInTheDocument();
    expect(screen.getAllByTestId('mock-date').length).toBe(31);
  });
});
