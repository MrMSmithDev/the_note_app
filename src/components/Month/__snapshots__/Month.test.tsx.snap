// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month component renders correctly to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col mb-2 md:mb-0"
  >
    <h3
      class="text-center font-bold font-quicksand text-md md:text-sm select-none dark:text-gray-100"
    >
      FEB
    </h3>
    <div
      class="grid grid-cols-7 gap-2 my-3 font-quicksand"
    >
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        S
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        M
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        T
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        W
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        T
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        F
      </span>
      <span
        class="text-md md:text-xs font-bold select-none dark:text-gray-200 text-center"
      >
        S
      </span>
      <div
        role="presentation"
        style="grid-column: span 3;"
      />
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-1
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-2
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-3
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-4
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-5
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-6
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-7
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-8
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-9
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-10
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-11
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-12
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-13
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-14
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-15
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-16
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-17
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-18
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-19
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-20
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-21
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-22
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-23
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-24
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-25
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-26
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-27
        </div>
      </div>
      <div
        class="font-quicksand text-center"
      >
        <div
          data-testid="mock-date"
        >
          Date for: 2025-2-28
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
